implementations/rust/vault/src/types.rs

```diff
@@ -0,0 +1,171 @@
+from_int_impl!(SecretKeyType, i32);
+from_int_impl!(SecretKeyType, i64);
+from_int_impl!(SecretKeyType, i128);
+from_int_impl!(SecretKeyType, u8);
+from_int_impl!(SecretKeyType, u16);
+from_int_impl!(SecretKeyType, u32);
+from_int_impl!(SecretKeyType, u64);
+from_int_impl!(SecretKeyType, u128);
+
+/// Persistence allowed by Secrets
+#[derive(Copy, Clone, Debug)]
+pub enum SecretPersistenceType {
+    /// Secret is temporary
+    Ephemeral,
+    /// Secret is permanent
+    Persistent,
+}
+
+impl SecretPersistenceType {
+    /// Convert enum to a number
+    pub fn to_usize(&self) -> usize {
+        match *self {
+            SecretPersistenceType::Ephemeral => 0,
+            SecretPersistenceType::Persistent => 1,
+        }
+    }
+}
+
+from_int_impl!(SecretPersistenceType, i8);
+from_int_impl!(SecretPersistenceType, i16);
+from_int_impl!(SecretPersistenceType, i32);
+from_int_impl!(SecretPersistenceType, i64);
+from_int_impl!(SecretPersistenceType, i128);
+from_int_impl!(SecretPersistenceType, u8);
+from_int_impl!(SecretPersistenceType, u16);
+from_int_impl!(SecretPersistenceType, u32);
+from_int_impl!(SecretPersistenceType, u64);
+from_int_impl!(SecretPersistenceType, u128);
+
+/// Secrets specific purpose
+#[derive(Copy, Clone, Debug)]
+pub enum SecretPurposeType {
+    /// Key exchange
+    KeyAgreement,
+}
+
+impl SecretPurposeType {
+    /// Convert enum to a number
+    pub fn to_usize(&self) -> usize {
+        match *self {
+            SecretPurposeType::KeyAgreement => 0,
+        }
+    }
+}
+
+from_int_impl!(SecretPurposeType, i8);
+from_int_impl!(SecretPurposeType, i16);
+from_int_impl!(SecretPurposeType, i32);
+from_int_impl!(SecretPurposeType, i64);
+from_int_impl!(SecretPurposeType, i128);
+from_int_impl!(SecretPurposeType, u8);
+from_int_impl!(SecretPurposeType, u16);
+from_int_impl!(SecretPurposeType, u32);
+from_int_impl!(SecretPurposeType, u64);
+from_int_impl!(SecretPurposeType, u128);
+
+/// Attributes for a specific vault secret
+#[derive(Copy, Clone, Debug)]
+pub struct SecretKeyAttributes {
+    /// The type of key
+    pub xtype: SecretKeyType,
+    /// How the key is persisted
+    pub persistence: SecretPersistenceType,
+    /// The purpose of the secret key
+    pub purpose: SecretPurposeType,
+}
+
+/// A context that uses secret keys e.g. TEE, HSM, SEP.
+/// This list is not meant to be exhaustive, just the ones supported
+/// Ockam vault.
+///
+/// Future options include:
+///
+/// Key is backed by Amazon's CloudHSM
+/// AmazonCloudHsm,
+/// Key is backed by Amazon's KMS
+/// AmazonKMS,
+/// Key is backed by Box's Keysafe
+/// BoxKeysafe,
+/// Key is backed HashiCorp KeyVault
+/// HashiCorpVault,
+/// Key is backed by an Intel SGX enclave
+/// IntelSgx,
+/// Key is backed by an AMD's PSP enclave
+/// AmdPsp,
+/// Key is backed by an ARM's Trustzone enclave
+/// ArmTrustzone,
+/// Key is backed a Apple's secure enclave
+/// SecureEnclave,
+#[derive(Clone, Copy, Debug)]
+pub enum SecretKeyContext {
```

Comment:
Would exposing type of vault/secret here create coupling we don't want?
For instance, we want someone to be able to plug-in a custom vault implementation without having to change Ockam code.

Reply:
In such cases to allow flexible usage it's convenient to use string as a type (or a subtype). Also, we can provide constants for some predefined types

Reply:
Context is where the key is and how is addressed.